book/src/simple-payment-and-state-verification.md

```diff
@@ -52,61 +52,43 @@ A receipt is a minimal proof that; a transaction has been included in a block,
+using Solana's fork selection rules.
 
-### Payment Merkle Path
+An Entry-Merkle is a Merkle Root including all transactions in a given entry,
+sorted by signature.
 
-A payment receipt is a data structure that contains a Merkle Path from a
-transaction to the required set of validator votes.
-
-An Entry-Merkle is a Merkle Root including all transactions in the entry, sorted
-by signature.
+A Block-Merkle is the Merkle Root of all the Entry-Merkles sequenced in the block.
 
 <img alt="Block Merkle Diagram" src="img/spv-block-merkle.svg" class="center"/>
 
-A Block-Merkle is a Merkle root of all the Entry-Merkles sequenced in the block.
+A Bank-Hash is the hash of the concatenation of the Block-Merkle and Accounts-Hash
+
+<img alt="Bank Hash Diagram" src="img/spv-bank-hash.svg" class="center"/>
+
+An Accounts-Hash is the hash of the concatentation of the state hashes of each
+account modified during the current slot.
+
 Transaction status is necessary for the receipt because the state receipt is
 constructed for the block. Two transactions over the same state can appear in
 the block, and therefore, there is no way to infer from just the state whether a
 transaction that is committed to the ledger has succeeded or failed in modifying
 the intended state. It may not be necessary to encode the full status code, but
 a single status bit to indicate the transaction's success.
 
-### State Merkle Path
-
-A state receipt provides a confirmation that a specific state is committed at the
-end of the block. Inter-block state transitions do not generate a receipt.
-
-For example:
+### Account State Verification
 
-* A sends 5 Lamports to B
-* B spends 5 Lamports
-* C sends 5 Lamports to A
-
-At the end of the block, A and B are in the exact same starting state, and any
-state receipt would point to the same value for A or B.
-
-The Bank-Merkle is computed from the Merkle Tree of the new state changes, along
-with the Previous Bank-Merkle, and the Block-Merkle.
-
-<img alt="Bank Merkle Diagram" src="img/spv-bank-merkle.svg" class="center"/>
-
-A state receipt contains only the state changes occurring in the block. A direct
-Merkle Path to the current Bank-Merkle guarantees the state value at that bank
-hash, but it cannot be used to generate a “current” receipt to the latest state
-if the state modification occurred in some previous block. There is no guarantee
-that the path provided by the validator is the latest one available out of all
-the previous Bank-Merkles.
-
-Clients that want to query the chain for a receipt of the "latest" state would
-need to create a transaction that would update the Merkle Path for that account,
-such as a credit of 0 Lamports.
+An account's state (balance or other data) can be verified by submitting a
+transaction with a ___TBD___ Instruction to the cluster. This instruction contains
+the slot boundary of interest and the expected state hash at that boundary. The
```

Comment:
We can’t enforce the slot boundary for this query.  I think it’s basically is the query true when it’s executed by the cluster.

Reply:
Hrm.  That's a pretty narrow window to hit on a high throughput account, basically the next slot and :crossed_fingers:, right?

Reply:
to be able to do "slot of interest":  would it be terribly expensive to hang onto the account hash vector at the end of every slot for 256 slots?  worst case is 256*2.5*sizeof(Hash)*TPS, similar to StatusCache

Reply:
@t-nelson why would the slot boundary provide any safety?  The state can change immediately on the next slot.  It’s up to the contract to have some meaningful terminal state, like “game over”

Reply:
@rob-solana by "hash vector" you mean the "inclusion proof" or "merkle path"?  They'd be log(N) and mergable per-slot, so that's certainly something to explore

@aeyakovenko there's no safety at the boundary.  I was getting at the race between receiving the "Don't squash this slot" instruction and the slot being squashed, though after review Carl's snapshot work, I realize we have until the bank in question falls out of the locktower.  This will require us scanning all bank forks and taking action on an unfinalized transaction, right?

Reply:
@t-nelson we can take action on it after that bank is rooted, and provide the result way later, maybe 1024 slots after or something long enough, or at an epoch boundary.